File: modules/Engine/src/Engine/GameState/Menu.cpp
#include <Engine/GameState/GameState.hpp>
#include <Engine/Resources.hpp>
#include <Engine/Settings.hpp>

#include <Util/Observer.hpp>


namespace engine::GameState
{

Menu::Menu()
: text("Menu, press Enter to continue,\nEsc to exit.", *Resources::get<sf::Font>("unifont"))
{
        this->text.setPosition(sf::Vector2f(Settings::Video::resolution) / 2.f);
        this->text.setOrigin({ this->text.getGlobalBounds().width  / 2.f,
                               this->text.getGlobalBounds().height / 2.f });
}

void Menu::handleInput(const sf::Event& event)
{
        if (event.type == sf::Event::KeyPressed)
        {
                switch (event.key.code)
                {
                        case sf::Keyboard::Return:
                        {
                                util::Subject::send({ util::Message::PushState{ new Game() }});
                                break;
                        }

                        case sf::Keyboard::Escape:
                        {
                                util::Subject::send({ util::Message::PopState{}});
                                break;
                        }

                        default: break;
                }
        }
}

void Menu::update()
{

}

void Menu::draw(sf::RenderTarget& target)
{
        target.draw(this->text);
}

}

File: data/script/GameState/Menu.lua
Menu = {
    drawables = {
        {
            type = "sfText",
            content = "Menu, press Enter to continue,\nEsc to exit.",
            font = "unifont",
            position = {
                x = Settings.video.resolution.x / 2,
                y = Settings.video.resolution.y / 2,
            },
            fillColor = "yellow",
            origin = "center"
        },
        {
            type = "sfCircleShape",
            fillColor = "yellow",
            radius = 30,
            position = { 40, 140 }
        },
        -- {
        --     type = "sfRectangleShape",
        --     fillColor = "blue",
        --     size = { 69, 42 },
        --     position = {
        --         Settings.video.resolution.x - 80,
        --         Settings.video.resolution.y - 50
        --     }
        -- },
    }
}

function Menu:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

function Menu:handle_input(event)
    if event.type == "KeyPressed" then
        local kc = event.code
        if kc == Keyboard.Enter then
            push_state "Game"
        elseif kc == Keyboard.Escape then
            pop_state()
        end
    end
end

function Menu:update(dt)
end

function Menu:draw()
    return self.drawables
end